### ISSUE TITLE
byValues does not guarantee the sort order

### ISSUE THREAD

**Issue by 0x009922** [cosimochellini/sort-es#6]
## Problem

Your documentation says:
> The sorter allows you to declare which properties are needed to sort the array, in **what order**, and with what sortable

And there is an example below:

```js
//i sort by THEIR NAMES and THEN by their ids
const sortedObject = objsToSort.sort(byValues({
    name: byString(),
    id: byNumber()
}));

// -- snip --

//i sort by THEIR IDS and THEN by their names
const sortedObject2 = objsToSort.sort(byValues({
    id: byNumber(),
    name: byString(),
}));
```

This actually works... but not in all cases. See this code example:

```js
console.log('String keys');

const ARR = [
    {
        id: 1,
        name: 'b',
    },
    {
        id: 2,
        name: 'a',
    },
    {
        id: 3,
        name: 'c',
    },
];

console.log(
    [...ARR].sort(
        byValues({
            id: byNumber(),
            name: byString(),
        }),
    ),
);
console.log(
    [...ARR].sort(
        byValues({
            name: byString(),
            id: byNumber(),
        }),
    ),
);

// Numeric keys - NOT OK!

console.log('\nNumeric keys');

const ARR_NUM = [
    {
        1: 1,
        2: 'b',
    },
    {
        1: 2,
        2: 'a',
    },
    {
        1: 3,
        2: 'c',
    },
];

console.log(
    [...ARR_NUM].sort(
        byValues({
            1: byNumber(),
            2: byString(),
        }),
    ),
);
console.log(
    [...ARR_NUM].sort(
        byValues({
            2: byString(),
            1: byNumber(),
        }),
    ),
);
```

This code produces this output:

```
String keys
[ { id: 1, name: 'b' }, { id: 2, name: 'a' }, { id: 3, name: 'c' } ]
[ { id: 2, name: 'a' }, { id: 1, name: 'b' }, { id: 3, name: 'c' } ]

Numeric keys
[ { '1': 1, '2': 'b' }, { '1': 2, '2': 'a' }, { '1': 3, '2': 'c' } ]
[ { '1': 1, '2': 'b' }, { '1': 2, '2': 'a' }, { '1': 3, '2': 'c' } ]
```

As you can see, in the second case, **the sort order has not changed**.

## Why?..

See [this](https://2ality.com/2015/10/property-traversal-order-es6.html#traversing-the-own-keys-of-an-object):
> Property keys are traversed in the following order:
> 
> - First, the keys that are integer indices (what these are is explained later), in ascending numeric order.
> - Then, all other string keys, in the order in which they were added to the object.
> - Lastly, all symbol keys, in the order in which they were added to the object.

Traversing keys in JavaScript objects does not guarantee the sequence in which these keys were added (in **ALL** cases). This works for `Map`, but not for plain objects.

## Option to solve the problem

You cannot guarantee the sort order through object keys. Creating a Map for this is too cumbersome. But you can sequentially specify fields for sorting in this form:

```js
[
    { id: 'asdf', 1: 4123, 0: false },
    { id: 'asdf', 1: 1232, 0: true },
    { id: 'asdf', 1: 5515, 0: false },
].sort(byValues(['id', byString()], [1, byNumber()]))
```

In this case, a list of arguments is passed from tuples `[number | string | symbol, sortable]`.

**Comment by cosimochellini** [comment]
Amazing, I was not aware of this lack of property sorting in JavaScript. 
Thank you so much for the meticulous issue you wrote to me, it's not something that happens often.
I had already initially thought about implementing byValues ​​as you suggested, but I had preferred to do it in the method currently for a nicer and more readable syntax. 
Consequently, I believe I will add an override to the method, avoiding breaking changes and allowing a user who has no special needs like yours, to be able to correctly use the sort with a nicer syntax.
It will be written and documented that in case of ordering of particular cases it is necessary to use the array syntax to be sure that the library will do its work.
Let me know what you think and if it can be a fair compromise, thanks again for the issue.

**Comment by cosimochellini** [comment]
@0x009922 Hi! I created the initial structure for your problem, [this is the branch](/cosimochellini/sort-es/tree/fix/6_byValues_incorrect_sorting).

Let me know if it is a suitable solution in your opinion or if you need anything else!

keep in mind that I will update the documentation and the description of the method to warn users that in case of problems the new method can be used to make sure that the array is sorted correctly.
😀

**Comment by 0x009922** [comment]
@cosimochellini, Hey!

It seems to me that it is better to get rid of `byValues` because its behavior and its documentation could be confusing to users. For now, of course, it's best to leave
it is completely unchanged, but mark it as `@deprecated`. In the documentation, also leave with a big red notice that "the method is outdated, it is better to use another", and then add some new method to the library, for example `byMultipleValues`, and redirect users to it.

How `@deprecated` looks at VS Code:

![image](https://user-images.githubusercontent.com/43530070/107321368-6e784c80-6ab3-11eb-982e-6d8e41535bd5.png)

**Comment by cosimochellini** [comment]
@0x009922  Hi!, actually thinking about it you are right, and I have seen that it is possible to deprecate even a specific signature of the method, resulting in a win for everyone.
![image](https://user-images.githubusercontent.com/24718181/107845362-2c187d80-6ddb-11eb-8139-a7f562a49f48.png)
![image](https://user-images.githubusercontent.com/24718181/107845377-3e92b700-6ddb-11eb-9a3b-27d4cc0e4acf.png)

One problem, I'm having a hard time making TS digest that there may also be a Symbol as the first parameter of the array, could you help me out?

The problem is the `SortableTuple<T>` in `types.ts` type.

**Comment by 0x009922** [comment]
Hi, @cosimochellini!

What's wrong with symbols? It works fine:

```ts
const id = Symbol('id')

const arr = [
  {
    [id]: 5
  }
];

arr.sort(byValues([[id, byNumber()]]))
```

Also, i think you should replace your's `byValues` typings:

```ts
function byValues<T>(sorters: SortableTuple<T>[]): sortable<T>;

/**
 * @deprecated do not pass the sorter for the byValue as object, use the `sorter: SortableTuple<T>[]`
 */
function byValues<T>(sorter: SortableObject<T>): sortable<T>;
```

...because when you type in IDE `byValues`, firstly it suggests first overload of method, which is deprecated. It may confuse.

**Comment by cosimochellini** [comment]
hi @0x009922 , I should have released version **1.2.0** which fixes the issue, let me know in case there are any problems!
